### PR TITLE
[agw][new feature] Idle mode support for multiple pdns and multiple dedicated bearers

### PR DESCRIPTION
**Pull request by pruthvihebbani**
[agw][new feature] Idle mode support for multiple pdns and multiple dedicated bearers

## Summary

This PR has code to re-establish the PDN/s and dedicated bearer/s after UE comes back from idle mode

## Test Plan
1. Verified S1 SIM sanity
2. Added a new TC test_attach_service_with_multi_pdns_and_bearers.py PR #2521 to verify the feature

### CODE REVIEW

File: lte/gateway/c/oai/tasks/sgw/sgw_handlers.c [hunk 1]
```diff
@@ -685,145 +726,176 @@ int sgw_handle_sgi_endpoint_deleted(
...
+
+          OAILOG_DEBUG_UE(
+              LOG_SPGW_APP, imsi64, "Delete GTPv1-U tunnel for sgw_teid : %d\n",
+              eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up);
+          // delete GTPv1-U tunnel
+          rv = gtp_tunnel_ops->del_tunnel(
+              ue, eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up,
+              eps_bearer_ctxt_p->enb_teid_S1u, NULL);
+          // This is best effort, ignore return code.
+          gtp_tunnel_ops->send_end_marker(enb, modify_bearer_pP->teid);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_handlers.c]: The end marker should be sent before the tunnel rule is deleted. Why did the ordering change here?

File: lte/gateway/c/oai/tasks/sgw/sgw_handlers.c [hunk 2]
```diff
@@ -685,145 +726,176 @@ int sgw_handle_sgi_endpoint_deleted(
   OAILOG_FUNC_RETURN(LOG_SPGW_APP, rv);
 }
 
+//------------------------------------------------------------------------------
+static void populate_sgi_end_point_update(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_handlers.c:840]: Could you put a comment about the function? Also in the signature, it is better to put the arguments to be modified in the function call at the end for readability.

File: lte/gateway/c/oai/tasks/sgw/sgw_handlers.c [hunk 3]
```diff
@@ -685,145 +726,176 @@ int sgw_handle_sgi_endpoint_deleted(
...
+    sgw_eps_bearer_ctxt_t* eps_bearer_ctxt_p,
+    const itti_s11_modify_bearer_request_t* const modify_bearer_pP,
+    itti_sgi_update_end_point_response_t* sgi_update_end_point_resp,
+    uint8_t sgi_rsp_idx, uint8_t* idx, imsi64_t imsi64) {
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_handlers.c]: is there a reason why idx is passed as a pointer and not passed by value as uint8_t?

File: lte/gateway/c/oai/tasks/sgw/sgw_handlers.c [hunk 4]
```diff
@@ -685,145 +726,176 @@ int sgw_handle_sgi_endpoint_deleted(
...
-        modify_bearer_pP->teid);
-
-    message_p->ittiMsgHeader.imsi = imsi64;
-    rv = send_msg_to_task(&spgw_app_task_zmq_ctx, TASK_MME, message_p);
-
     OAILOG_FUNC_RETURN(LOG_SPGW_APP, rv);
   }
-
-  OAILOG_FUNC_RETURN(LOG_SPGW_APP, RETURNerror);
+  sgw_handle_sgi_endpoint_updated(&sgi_update_end_point_resp, imsi64);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_handlers.c]: move this to the end of the `if (bearer_ctxt_info_p) { ` to make it more readable as this is only supposed to be done when the if statement is true.

File: lte/gateway/c/oai/tasks/sgw/sgw_handlers.c [hunk 5]
```diff
@@ -685,145 +726,176 @@ int sgw_handle_sgi_endpoint_deleted(
...
-    modify_response_p->trxn              = modify_bearer_pP->trxn;
-    OAILOG_DEBUG_UE(
-        LOG_SPGW_APP, imsi64,
-        "Rx MODIFY_BEARER_REQUEST, teid " TEID_FMT " CONTEXT_NOT_FOUND\n",
-        modify_bearer_pP->teid);
-
-    message_p->ittiMsgHeader.imsi = imsi64;
-    rv = send_msg_to_task(&spgw_app_task_zmq_ctx, TASK_MME, message_p);
-
     OAILOG_FUNC_RETURN(LOG_SPGW_APP, rv);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_handlers.c]: can be handled by line 898 once line 897 is moved into the if clause....

**Review by ulaskozat — APPROVED** [review]
LGTM

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 6]
```diff
@@ -1193,26 +1193,285 @@ int mme_app_handle_create_sess_resp(
...
+            ue_context_p,
+            initial_ctxt_setup_rsp_p->e_rab_setup_list.item[item].e_rab_id)) ==
+        NULL) {
+      s11_modify_bearer_request->bearer_contexts_to_be_removed
+          .bearer_contexts[(*idx)++]
+          .eps_bearer_id =
+          initial_ctxt_setup_rsp_p->e_rab_setup_list.item[item].e_rab_id;
+    }
+  }
+  OAILOG_FUNC_OUT(LOG_MME_APP);
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:1214]: Also add a corresponding OAILOG_FUNC_IN call at the beginning of the function

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 7]
```diff
@@ -1193,26 +1193,285 @@ int mme_app_handle_create_sess_resp(
...
+            "Failed to release the dedicated EPS bearer context for "
+            "ebi:%u\n",
+            initial_ctxt_setup_rsp_p->e_rab_failed_to_setup_list.item[item]
+                .e_rab_id);
+      }
+      break;
+    }  // end of ebi comparison
+  }    // end of for
+
+  OAILOG_FUNC_OUT(LOG_MME_APP);
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:1263]: Add OAILOG_FUNC_IN too at the beginning of the function

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 8]
```diff
@@ -1193,26 +1193,285 @@ int mme_app_handle_create_sess_resp(
...
+}
+//------------------------------------------------------------------------------
+static void mme_app_populate_bearer_contexts_to_be_modified(
+    struct ue_mm_context_s* ue_context_p,
+    itti_mme_app_initial_context_setup_rsp_t* const initial_ctxt_setup_rsp_p,
+    itti_s11_modify_bearer_request_t* s11_modify_bearer_request, uint8_t pid,
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c]: Remove the "pid" argument as it is not used in this function.

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 9]
```diff
@@ -1193,26 +1193,285 @@ int mme_app_handle_create_sess_resp(
...
+
+  s11_modify_bearer_request->teid =
+      ue_context_p->pdn_contexts[*pid]->s_gw_teid_s11_s4;
+
+  s11_modify_bearer_request->bearer_contexts_to_be_modified.num_bearer_context =
+      bc_to_be_modified_idx;
+  s11_modify_bearer_request->bearer_contexts_to_be_removed.num_bearer_context =
+      *bc_to_be_removed_idx;
+  s11_modify_bearer_request->mme_fq_csid.node_id_type =
+      GLOBAL_UNICAST_IPv4;                          // TODO
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:1387]: Add little more context on what is TODO here

**Review by ulaskozat — CHANGES_REQUESTED** [review]
When tested with real UE and eNB, the bearers that are re-established do not have the correct TFT rules and priorities installed.

**Review by ulaskozat — APPROVED** [review]

**Pull request merged**